Open-search process crashing on using the index.lifecycle.name in the API body #102

**Describe the bug**
While applying the policy _index_template on an index, Open-search crashed on using the index.lifecycle.name in the API body. This setting is not available in open-search but it shouldn't crash the process and return with the exception.

This is occurring on a vanilla OS cluster even without index management plugin installed on the cluster. 

Issue opened in Index Management Repo: https://github.com/opensearch-project/index-management/issues/102

**To Reproduce**
* Run an open-search cluster. 
* Hit the endpoint with an API: `curl -X PUT "localhost:9200/_index_template/timeseries_template?pretty" -H 'Content-Type: application/json' -d'
{
  "index_patterns": ["*"],
  "data_stream": { },
  "template": {
    "settings": {
      "number_of_shards": 1,
      "number_of_replicas": 1,
      "index.lifecycle.name": "timeseries_policy"
    }
  }
}
'
`
* The response returned is `curl: (52) Empty reply from server`
* The log trace found in the ES logs is pasted below. 

**Expected behavior**
As the setting is not available the process shouldn't crash but return the exception as setting not found.

**Log Trace**
`
=== Standard error of node `node{::runTask-0}` ===
»   ↓ last 40 non error or warning messages from /local/home/aditjind/gh/OpenSearch/build/testclusters/runTask-0/logs/opensearch.stderr.log ↓
» fatal error in thread [opensearch[runTask-0][masterService#updateTask][T#1]], exiting
»  java.lang.AssertionError: expected [index.lifecycle.name] to be private but it was not
»       at org.opensearch.cluster.metadata.MetadataCreateIndexService.validatePrivateSettingsNotExplicitlySet(MetadataCreateIndexService.java:1047)
»       at org.opensearch.cluster.metadata.MetadataCreateIndexService.getIndexSettingsValidationErrors(MetadataCreateIndexService.java:1037)
»       at org.opensearch.cluster.metadata.MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:1305)
»       at org.opensearch.cluster.metadata.MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:1226)
»       at org.opensearch.cluster.metadata.MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:1221)
»       at org.opensearch.cluster.metadata.MetadataIndexTemplateService.addIndexTemplateV2(MetadataIndexTemplateService.java:507)
»       at org.opensearch.cluster.metadata.MetadataIndexTemplateService$4.execute(MetadataIndexTemplateService.java:403)
»       at org.opensearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:60)
»       at org.opensearch.cluster.service.MasterService.executeTasks(MasterService.java:715)
»       at org.opensearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:337)
»       at org.opensearch.cluster.service.MasterService.runTasks(MasterService.java:232)
»       at org.opensearch.cluster.service.MasterService.access$000(MasterService.java:86)
»       at org.opensearch.cluster.service.MasterService$Batcher.run(MasterService.java:164)
»       at org.opensearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:163)
»       at org.opensearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:201)
»       at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:697)
»       at org.opensearch.common.util.concurrent.PrioritizedOpenSearchThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedOpenSearchThreadPoolExecutor.java:266)
»       at org.opensearch.common.util.concurrent.PrioritizedOpenSearchThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedOpenSearchThreadPoolExecutor.java:229)
»       at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
»       at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
»       at java.base/java.lang.Thread.run(Thread.java:832)`

**Host/Environment (please complete the following information):**
 - OS: Linux
 - Version: 5.4.116-64.217.amzn2int.x86_64

--- a/server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java
+++ b/server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java
@@ -1044,7 +1044,10 @@ public class MetadataCreateIndexService {
         for (final String key : settings.keySet()) {
             final Setting<?> setting = indexScopedSettings.get(key);
             if (setting == null) {
-                assert indexScopedSettings.isPrivateSetting(key) : "expected [" + key + "] to be private but it was not";
+                //https://github.com/opensearch-project/OpenSearch/issues/1019
+                if(!indexScopedSettings.isPrivateSetting(key)) {
+                    validationErrors.add("expected [" + key + "] to be private but it was not");
+                }
             } else if (setting.isPrivateIndex()) {
                 validationErrors.add("private index setting [" + key + "] can not be set explicitly");
             }

--- a/server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java
+++ b/server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java
@@ -1044,7 +1044,7 @@ public class MetadataCreateIndexService {
         for (final String key : settings.keySet()) {
             final Setting<?> setting = indexScopedSettings.get(key);
             if (setting == null) {
-                //https://github.com/opensearch-project/OpenSearch/issues/1019
+                // see: https://github.com/opensearch-project/OpenSearch/issues/1019
                 if(!indexScopedSettings.isPrivateSetting(key)) {
                     validationErrors.add("expected [" + key + "] to be private but it was not");
                 }

--- a/server/src/test/java/org/opensearch/cluster/metadata/MetadataCreateIndexServiceTests.java
+++ b/server/src/test/java/org/opensearch/cluster/metadata/MetadataCreateIndexServiceTests.java
@@ -965,7 +965,7 @@ public class MetadataCreateIndexServiceTests extends OpenSearchTestCase {
     }
 
     public void testIndexLifecycleNameSetting() {
-        //https://github.com/opensearch-project/OpenSearch/issues/1019
+        // see: https://github.com/opensearch-project/OpenSearch/issues/1019
         final Settings ilnSetting = Settings.builder().put("index.lifecycle.name", "dummy").build();
         withTemporaryClusterService(((clusterService, threadPool) -> {
             MetadataCreateIndexService checkerService = new MetadataCreateIndexService(

--- a/server/src/test/java/org/opensearch/cluster/metadata/MetadataCreateIndexServiceTests.java
+++ b/server/src/test/java/org/opensearch/cluster/metadata/MetadataCreateIndexServiceTests.java
@@ -66,6 +66,7 @@ import org.opensearch.common.unit.TimeValue;
 import org.opensearch.common.util.BigArrays;
 import org.opensearch.common.xcontent.NamedXContentRegistry;
 import org.opensearch.common.xcontent.XContentFactory;
+import org.opensearch.env.Environment;
 import org.opensearch.index.Index;
 import org.opensearch.index.IndexNotFoundException;
 import org.opensearch.index.IndexSettings;
@@ -961,6 +962,31 @@ public class MetadataCreateIndexServiceTests extends OpenSearchTestCase {
             Collections.emptySet());
         assertWarnings("Translog retention settings [index.translog.retention.age] "
             + "and [index.translog.retention.size] are deprecated and effectively ignored. They will be removed in a future version.");
+    }
+
+    public void testIndexLifecycleNameSetting() {
+        //https://github.com/opensearch-project/OpenSearch/issues/1019
+        final Settings ilnSetting = Settings.builder().put("index.lifecycle.name", "dummy").build();
+        withTemporaryClusterService(((clusterService, threadPool) -> {
+            MetadataCreateIndexService checkerService = new MetadataCreateIndexService(
+                Settings.EMPTY,
+                clusterService,
+                null,
+                null,
+                null,
+                createTestShardLimitService(randomIntBetween(1, 1000), clusterService),
+                new Environment(Settings.builder().put("path.home", "dummy").build(), null),
+                new IndexScopedSettings(ilnSetting, Collections.emptySet()),
+                threadPool,
+                null,
+                new SystemIndices(Collections.emptyMap()),
+                true
+            );
+
+            final List<String> validationErrors =  checkerService.getIndexSettingsValidationErrors(ilnSetting, true);
+            assertThat(validationErrors.size(), is(1));
+            assertThat(validationErrors.get(0), is("expected [index.lifecycle.name] to be private but it was not"));
+        }));
     }
 
     private IndexTemplateMetadata addMatchingTemplate(Consumer<IndexTemplateMetadata.Builder> configurator) {

